keras_cv/losses/dice.py

```diff
@@ -0,0 +1,215 @@
+                `generalized` or `adaptive`; otherwise compute original dice score.
+                Default to `None`.
+            label_smoothing: Float in [0, 1]. When > 0, label values are smoothed,
+                meaning the confidence on label values are relaxed. For example, if
+                `0.1`, use `0.1 / num_classes` for non-target labels and
+                `0.9 + 0.1 / num_classes` for target labels. Default to `0.0`.
+            per_sample: If `True`, the loss will be calculated for each sample in
+                batch and then averaged. Otherwise the loss will be calculated for
+                the whole batch. Default to `False`.
+            epsilon: Small float added to dice score to avoid dividing by zero.
+                Default to `1e-07`.
+            name: Optional name for the instance.
+                Defaults to 'dice'.
+        """
+        super().__init__(name=name, **kwargs)
+        self.beta = beta
+        self.from_logits = from_logits
+        self.loss_type = loss_type
+        self.label_smoothing = label_smoothing
+        self.per_sample = per_sample
+        self.epsilon = epsilon
+
+        self.axis = tf.constant(axis)
+        if not self.per_sample:
+            self.axis = tf.concat([tf.constant([0]), self.axis], axis=0)
+
+        if class_ids is not None:
+            if isinstance(class_ids, float) or any(
+                isinstance(x, float) for x in class_ids
+            ):
+                raise ValueError(
+                    f"The indices should be int or a list of integer. Got {class_ids}"
+                )
+            elif isinstance(class_ids, int):
+                class_ids = [class_ids]
+
+        if self.loss_type is not None:
+            if self.loss_type not in ["generalized", "adaptive"]:
+                raise ValueError(
+                    "The loss type is not valid. "
+                    f"It should be `generalized` or `adaptive`. Got {self.loss_type}"
+                )
+
+        self.class_ids = class_ids
+
+    def _smooth_labels(self, y_true, y_pred, label_smoothing):
+        num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
+        return y_true * (1.0 - label_smoothing) + (label_smoothing / num_classes)
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
+        label_smoothing = tf.convert_to_tensor(self.label_smoothing, dtype=y_pred.dtype)
+
+        if self.from_logits:
+            y_pred = tf.nn.softmax(y_pred)
+
+        if tf.cast(label_smoothing, dtype=tf.bool):
+            y_true = self._smooth_labels(y_true, y_pred, label_smoothing)
+
+        if self.class_ids is not None:
+            y_true, y_pred = gather_channels(y_true, y_pred, indices=self.class_ids)
+
+        # loss calculation: Fβ-score (in terms of Type I and type II erro
+        true_positive = keras.backend.sum(y_true * y_pred, axis=self.axis)
```

Comment:
I think we are hiding most of the keras.backend functions since they are just simple wrapper of tf ops. Please replace this with tf.math ops instead.

Reply:
Done.